CAMEL-12543: Add camel-debezium component to Camel (#3128)
apache/camel: components/camel-debezium/src/main/docs/debezium-component.adoc

[[debezium-component]]
= Debezium Component

*Available as of Camel version 3.0*

The Debezium component is wrapper around https://debezium.io/[Debezium] using https://debezium.io/docs/embedded/[Debezium Embedded], which enabled Change Data Capture from various databases without the need of Kafka or Kafka Connect.

> *Note on handling failures:* Per https://debezium.io/docs/embedded/#handling_failures[Debezium Embedded Engine] documentation, the engines is actively recording source offsets and periodically flushes these offsets to a persistent storage, so when the application is restarted or crashed, the engine will resume from the last recorded offset.

Reviewer: Should point to https://debezium.io/documentation/reference/0.9/operations/embedded.html (we just re-structured the docs). The old link will be redirected, but better to avoid it.

> Thus, at normal operation, your downstream routes will receive exactly once event, however in case of application crash (not having a graceful shutdown), the application will resume from the last recorded offset,

Reviewer: ```suggestion
Thus, at normal operation, your downstream routes will receive each event exactly once, however in case of an application crash (not having a graceful shutdown), the application will resume from the last recorded offset,
```

> which may result of receiving duplicate events immediately after the restart. Therefore, your downstream routes should be tolerant enough of such case and deduplicate events if needed.

Reviewer: ```suggestion
which may result in receiving duplicate events immediately after the restart. Therefore, your downstream routes should be tolerant enough of such case and deduplicate events if needed.
```

Maven users will need to add the following dependency to their `pom.xml`
for this component.

[source,xml]
------------------------------------------------------------
<dependency>
    <groupId>org.apache.camel</groupId>
    <artifactId>camel-debezium</artifactId>
    <version>x.x.x</version>
    <!-- use the same version as your Camel core version -->
</dependency>
------------------------------------------------------------

== URI format

[source,java]
---------------------------
debezium:connector-type[?options]

---------------------------

== Supported Debezium Connectors

> - https://debezium.io/docs/connectors/mysql/[MySql].

Reviewer: Ok, so if you *only* support MySQL for now, then the fixed dependency would be ok. I'd then add a sentence here that other connectors are not supported yet.

Author: Added

== Options

// component options: START
The Debezium component supports 2 options, which are listed below.

[width="100%",cols="2,5,^1,2",options="header"]
|===
| Name | Description | Default | Type
| *configuration* (consumer) | Allow pre-configured Configurations to be set, you will need to extend EmbeddedDebeziumConfiguration in order to create the configuration for the component |  | EmbeddedDebezium Configuration
| *basicPropertyBinding* (advanced) | Whether the component should use basic property binding (Camel 2.x) or the newer property binding with additional capabilities | false | boolean
|===
// component options: END

// endpoint options: START
The Debezium endpoint is configured using URI syntax:

----
debezium:connectorType
----

with the following path and query parameters:

=== Path Parameters (1 parameters):

[width="100%",cols="2,5,^1,2",options="header"]
|===
| Name | Description | Default | Type

> | *connectorType* | *Required* The Debezium connector type that is supported by Camel Debezium component. |  | String

Reviewer: Should this document the allowed values, i.e. "mysql" at this point?

Author: I can add that

| *offsetStorageReplication Factor* (consumer) | Replication factor used when creating the offset storage topic. Required when offset.storage is set to the KafkaOffsetBackingStore |  | int
| *offsetStorageTopic* (consumer) | The name of the Kafka topic where offsets are to be stored. Required when offset.storage is set to the KafkaOffsetBackingStore. |  | String
| *exceptionHandler* (consumer) | To let the consumer use a custom ExceptionHandler. Notice if the option bridgeErrorHandler is enabled then this option is not in use. By default the consumer will deal with exceptions, that will be logged at WARN or ERROR level and ignored. |  | ExceptionHandler
| *exchangePattern* (consumer) | Sets the exchange pattern when the consumer creates an exchange. |  | ExchangePattern
| *basicPropertyBinding* (advanced) | Whether the endpoint should use basic property binding (Camel 2.x) or the newer property binding with additional capabilities | false | boolean
| *synchronous* (advanced) | Sets whether synchronous processing should be strictly used, or Camel is allowed to use asynchronous processing (if supported). | false | boolean

> | *bigintUnsignedHandlingMode* (mysql) | Specifies how BIGINT UNSIGNED columns should be represented in change events, including: precise uses java.math.BigDecimal to represent values, which are encoded in the change events using a binary representation and Kafka Connects org.apache.kafka.connect.data.Decimal type; long (the default) represents values using Javas long, which may not offer the precision but will be far easier to use in consumers. long is usually the preferable setting. Only when working with values larger than 263, the precise setting should be used as those values cant be conveyed using long. See Data types. | long | String

Reviewer: I'd advise against commenting all the connector-specific options here, as this easily will become outdated. Instead refer to the upstream docs, e.g. https://debezium.io/documentation/reference/0.9/connectors/mysql.html#connector-properties

Author: Maybe we can change the explanation to something like: "Corresponding Debezium property configuration `decimal.handling.mode`. Refer to Debezium documentation for detailed property explanation" ?

|===
// endpoint options: END
// spring-boot-auto-configure options: START
== Spring Boot Auto-Configuration

When using Spring Boot make sure to use the following Maven dependency to have support for auto configuration:

[source,xml]
----
<dependency>
  <groupId>org.apache.camel</groupId>
  <artifactId>camel-debezium-starter</artifactId>
  <version>x.x.x</version>
  <!-- use the same version as your Camel core version -->
</dependency>
----

The component supports 3 options, which are listed below.

[width="100%",cols="2,5,^1,2",options="header"]
|===
| Name | Description | Default | Type
| *camel.component.debezium.basic-property-binding* | Whether the component should use basic property binding (Camel 2.x) or the newer property binding with additional capabilities | false | Boolean
| *camel.component.debezium.configuration* | Allow pre-configured Configurations to be set, you will need to extend EmbeddedDebeziumConfiguration in order to create the configuration for the component. The option is a org.apache.camel.component.debezium.configuration.EmbeddedDebeziumConfiguration type. |  | String

> | *camel.component.debezium.enabled* | Whether to enable auto configuration of the debezium component. This is enabled by default. |  | Boolean

Reviewer: ```suggestion
| *camel.component.debezium.enabled* | Whether to enable auto configuration of the Debezium component. This is enabled by default. |  | Boolean
```

|===
// spring-boot-auto-configure options: END

For more information about configuration:

> https://debezium.io/docs/embedded/#engine-properties[https://debezium.io/docs/embedded/#engine-properties]

Reviewer: Should be updated as per above.

https://debezium.io/docs/connectors/mysql/#connector-properties[https://debezium.io/docs/connectors/mysql/#connector-properties]

== Message headers

=== Consumer headers

> The following headers are available when consuming messages from Kafka.

Reviewer: ```suggestion
The following headers are available when consuming change events from Debezium.
```

[width="100%",cols="2m,2m,1m,5",options="header"]
|===
| Header constant                           | Header value                                   | Type        | Description
| DebeziumConstants.HEADER_IDENTIFIER       | "CamelDebeziumIdentifier"                      | String      | The identifier of the connector, normally is this format "{server-name}.{database-name}.{table-name}".
| DebeziumConstants.HEADER_KEY              | "CamelDebeziumKey"                             | Object      | The key of the event, normally is the table Primary Key.

> | DebeziumConstants.HEADER_SOURCE_METADATA  | "CamelDebeziumSourceMetadata"                  | Map         | The metadata about the source event, for example `table` name, database `name`, log position, etc, please refer to debezium documentation for more info.

Reviewer: ```suggestion
| DebeziumConstants.HEADER_SOURCE_METADATA  | "CamelDebeziumSourceMetadata"                  | Map         | The metadata about the source event, for example `table` name, database `name`, log position, etc, please refer to the Debezium documentation for more info.
```

> | DebeziumConstants.HEADER_OPERATION        | "CamelDebeziumOperation"                       | String      | If presents, the type of event operation. Values for the connector are c for create (or insert), u for update, d for delete.

Reviewer: or "r" in case of a snapshot event

| DebeziumConstants.HEADER_TIMESTAMP        | "CamelDebeziumTimestamp"                       | Long        | If presents, the time (using the system clock in the JVM) at which the connector processed the event.
| DebeziumConstants.HEADER_BEFORE           | "CamelDebeziumBefore"                          | Map/Struct  | If presents, contains the state of the row before the event occurred.
|===

== Samples

=== Consuming events

> Here is a very simple route that you can use in order to listen to Debezium events from MySql connector.

Reviewer: ```suggestion
Here is a very simple route that you can use in order to listen to Debezium events from MySQL connector.
```

[source,java]
----
from("debezium:mysql?name=dbz-test-1&offsetStorageFileName=/usr/offset-file-1.dat&databaseHostName=localhost&databaseUser=debezium&databasePassword=dbz&databaseServerName=my-app-connector&databaseHistoryFileName=/usr/history-file-1.dat")
    .log("Event received from Debezium : ${body}")
    .log("    with this identifier ${headers.CamelDebeziumIdentifier}")
    .log("    with these source metadata ${headers.CamelDebeziumSourceMetadata}")
    .log("    the event occured upon this operation '${headers.CamelDebeziumSourceOperation}'")
    .log("    on this database '${headers.CamelDebeziumSourceMetadata[db]}' and this table '${headers.CamelDebeziumSourceMetadata[table]}'")
    .log("    with the key ${headers.CamelDebeziumKey}")
    .log("    the previous value is ${headers.CamelDebeziumBefore}")
----

By default, the component will emit the events in the body and `CamelDebeziumBefore` header as https://kafka.apache.org/22/javadoc/org/apache/kafka/connect/data/Struct.html[`Struct`] data type, the reasoning behind this, is to perceive the schema information in case is needed.

> However, the component as well contains a https://camel.apache.org/manual/latest/type-converter.html[Type Converter] that converts from

Reviewer 2: ```suggestion
However, the component as well contains a xref:manual::type-converter.adoc[Type Converter] that converts from
```

Author: Noted and thanks a lot for the hints !

> from default output type of https://kafka.apache.org/22/javadoc/org/apache/kafka/connect/data/Struct.html[`Struct`] to `Map` in order to leverage Camel's rich https://camel.apache.org/manual/latest/data-format.html[Data Format] types which many of them work out of box with `Map` data type.

Reviewer 2: ```suggestion
from default output type of https://kafka.apache.org/22/javadoc/org/apache/kafka/connect/data/Struct.html[`Struct`] to `Map` in order to leverage Camel's rich xref:manual::data-format.adoc[Data Format] types which many of them work out of box with `Map` data type.
```

To use it, you can either add `Map.class` type when you access the message e.g: `exchange.getIn().getBody(Map.class)`, or you can convert the body always to `Map` from the route builder by adding `.convertBodyTo(Map.class)` to your Camel Route DSL after `from` statement.

We mentioned above about the schema, which can be used in case you need to perform advance data transformation and the schema is needed for that. If you choose not to convert your body to `Map`,
you can obtain the schema information as https://kafka.apache.org/22/javadoc/org/apache/kafka/connect/data/Schema.html[`Schema`] type from `Struct` like this:
[source,java]
----
from("debezium:[connectorType]?[options]])
    .process(exchange -> {
        final Struct bodyValue = exchange.getIn().getBody(Struct.class);
        final Schema schemaValue = bodyValue.schema();

        log.info("Body value is :" + bodyValue);
        log.info("With Schema : " + schemaValue);
        log.info("And fields of :" + schemaValue.fields());
        log.info("Field name has `" + schemaValue.field("name").schema() + "` type");
    });
----

> *Important Note:* This component is a thin wrapper around Debezium Engine as mentioned, therefore before using this component in production, you need to understand how Debezium works and how configurations can reflect the expected behavior, especially in regards to https://debezium.io/docs/embedded/#handling_failures[handling failures].

Reviewer: URL should be updated as per above.
